// uttp/dist/adapters/fetch/index.js
import {
  getUrlWithBase
} from "../../chunk-K66QKAHF.js";

// src/adapters/fetch/index.ts
var getFetchAdapter = (handler) => {
  return async (...options) => {
    const handlerBag = await handler({
      parseBodyAsString(rawRequest) {
        const request = rawRequest;
        return request.text();
      }
    }, ...options);
    return async (request) => {
      const url = getUrlWithBase(request.url);
      const response = await handlerBag.handleRequest({
        rawRequest: request,
        body: void 0,
        headers: Object.fromEntries(request.headers),
        method: request.method,
        url
      });
      const fetchResponse = new Response(response.body, {
        status: response.status
      });
      if (response.headers) {
        for (const [key, value] of Object.entries(response.headers)) {
          if (typeof value === "undefined")
            continue;
          if (typeof value === "string") {
            fetchResponse.headers.set(key, value);
            continue;
          }
          for (const v of value)
            fetchResponse.headers.append(key, v);
        }
      }
      return fetchResponse;
    };
  };
};
export {
  getFetchAdapter
};
