// uttp/dist/adapters/aws-lambda/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkSV7F6ZFDcjs = require('../../chunk-SV7F6ZFD.cjs');

// src/adapters/aws-lambda/index.ts
var getAwsLambdaAdapter = (handler) => {
  return async (...options) => {
    const handlerBag = await handler({
      parseBodyAsString(rawRequest) {
        const request = rawRequest;
        return request.body;
      }
    }, ...options);
    const requestHandler = async (event) => {
      var _a;
      const url = _chunkSV7F6ZFDcjs.getUrlWithBase.call(void 0, `${event.rawPath}?${event.rawQueryString}`);
      const response = await handlerBag.handleRequest({
        rawRequest: event,
        body: event.body,
        headers: event.headers,
        method: event.requestContext.http.method,
        url
      });
      const headers = {};
      if (response.headers) {
        for (const [key, value] of Object.entries(response.headers)) {
          if (typeof value === "undefined")
            continue;
          headers[key] = typeof value === "string" ? value : value.join(",");
        }
      }
      return {
        statusCode: response.status,
        body: (_a = response.body) == null ? void 0 : _a.toString(),
        headers
      };
    };
    return requestHandler;
  };
};


exports.getAwsLambdaAdapter = getAwsLambdaAdapter;
