// uttp/dist/adapters/koa/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkSV7F6ZFDcjs = require('../../chunk-SV7F6ZFD.cjs');

// src/adapters/koa/index.ts
var getKoaAdapter = (handler) => {
  return async (...options) => {
    const handlerBag = await handler({
      parseBodyAsString(rawRequest) {
        const request = rawRequest;
        return _chunkSV7F6ZFDcjs.getStringFromIncomingMessage.call(void 0, request, { maxBodySize: handlerBag.adapterOptions.maxBodySize });
      }
    }, ...options);
    const middleware = async (context) => {
      const response = await handlerBag.handleRequest({
        rawRequest: context.req,
        body: context.body,
        headers: context.headers,
        method: context.method,
        url: context.URL
      });
      context.status = response.status;
      if (response.headers) {
        for (const [key, value] of Object.entries(response.headers)) {
          if (typeof value === "undefined")
            continue;
          context.set(key, value);
        }
      }
      context.body = response.body;
    };
    return middleware;
  };
};


exports.getKoaAdapter = getKoaAdapter;
