// uttp/dist/chunk-YQ4ORDYY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _chunkSV7F6ZFDcjs = require('./chunk-SV7F6ZFD.cjs');

// src/adapters/node/index.ts
var getNodeAdapter = (handler) => {
  return async (...options) => {
    const handlerBag = await handler({
      parseBodyAsString(rawRequest) {
        const request = rawRequest;
        return _chunkSV7F6ZFDcjs.getStringFromIncomingMessage.call(void 0, request, { maxBodySize: handlerBag.adapterOptions.maxBodySize });
      }
    }, ...options);
    const requestListener = async (request, nodeResponse) => {
      const url = _chunkSV7F6ZFDcjs.getUrlWithBase.call(void 0, request.url);
      const response = await handlerBag.handleRequest({
        rawRequest: request,
        body: void 0,
        headers: request.headers,
        method: request.method,
        url
      });
      nodeResponse.statusCode = response.status;
      if (response.headers) {
        for (const [key, value] of Object.entries(response.headers)) {
          if (typeof value === "undefined")
            continue;
          nodeResponse.setHeader(key, value);
        }
      }
      nodeResponse.end(response.body);
    };
    return requestListener;
  };
};



exports.getNodeAdapter = getNodeAdapter;
